### ISSUE TITLE
importing fails when calling python -OO

### ISSUE THREAD

**Issue by keewis** [dask/dask#6296]
This was discovered by `xarray`'s `upstream-dev` CI ([environment](https://dev.azure.com/xarray/xarray/_build/results?buildId=2996&view=logs&j=2280efed-fda1-53bd-9213-1fa8ec9b4fa8&t=031ddd67-e55f-5fbd-2283-1ff4dfed6587)) a few days ago, but we were a bit slow in reporting so this also happens with the newly released `2.18.0`.

The problem is this:
```
$ python -OO -c 'import dask.array'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File ".../lib/python3.8/site-packages/dask/array/__init__.py", line 26, in <module>
    from .routines import (
  File ".../lib/python3.8/site-packages/dask/array/routines.py", line 18, in <module>
    from .creation import arange, diag, empty, indices
  File ".../lib/python3.8/site-packages/dask/array/creation.py", line 26, in <module>
    from .wrap import empty, ones, zeros, full
  File ".../lib/python3.8/site-packages/dask/array/wrap.py", line 173, in <module>
    full.__name__ = _full.__name__
AttributeError: 'functools.partial' object has no attribute '__name__'
```
without the optimization, the import obviously works.

See also pydata/xarray#4124

**Comment by jsignell** [comment]
Ping @dask/maintenance for visibility. I am taking a look now.

**Comment by jrbourbeau** [comment]
@keewis just as an FYI we released `dask==2.18.1` which contains @jsignell's fix from https://github.com/dask/dask/pull/6299. Thanks for reporting and thanks @jsignell for the fix!
